### PR TITLE
Improve ProjectBuildingException error messages with detailed problem reporting

### PR DESCRIPTION
**Pull request by PasanAbeysekara**
- https://github.com/apache/maven/issues/10951

## Description

Enhanced `ProjectBuildingException` to provide detailed error messages instead of the generic "Some problems were encountered while processing the POMs" message. This improvement helps developers quickly identify and fix build problems without needing to use a debugger.

## What was changed

- Modified the `ProjectBuildingException(List<ProjectBuildingResult> results)` constructor to generate comprehensive error messages
- Added a new `createMessage(List<ProjectBuildingResult> results)` method that:
  - Counts total problems and errors across all project results
  - Provides a summary with error count (e.g., "3 problems were encountered while processing the POMs (2 errors)")
  - Lists detailed problems for each project with severity, message, and location information
  - Shows precise source locations (file, line, column) when available
  - Prioritizes errors over warnings for cleaner output when both are present
  - Handles edge cases gracefully (missing project IDs, file paths, etc.)

## Outcome

**Before:**

```
Some problems were encountered while processing the POMs
```

**After:**

```
3 problems were encountered while processing the POMs (2 errors):

[com.example:project1:1.0] [ERROR] Missing required dependency @ pom.xml, line 25, column 10 [ERROR] Invalid version format @ pom.xml, line 30, column 5

[com.example:project2:1.0] [WARNING] Deprecated plugin usage @ pom.xml, line 15, column 3
```

This enhancement significantly improves the developer experience by providing actionable error information without requiring debugging tools.

## Checklist

Following this checklist to help us incorporate your contribution quickly and easily:

- [x] Your pull request should address just one issue, without pulling in other changes.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Each commit in the pull request should have a meaningful subject line and body.
  Note that commits might be squashed by a maintainer on merge.
- [ ] Write unit tests that match behavioral changes, where the tests fail if the changes to the runtime are not applied.
  This may not always be possible but is a best-practice.
- [ ] Run `mvn verify` to make sure basic checks pass.
  A more thorough check will be performed on your pull request automatically.
- [ ] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

- [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
- [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

## Notes

- The change is backward compatible and doesn't affect existing API contracts
- Error message format follows Maven's established conventions for problem reporting
- Implementation handles edge cases and null safety appropriately

### CODE REVIEW

File: impl/maven-core/src/test/java/org/apache/maven/project/ProjectBuildingExceptionMessageDemo.java [hunk 1]
```diff
@@ -0,0 +1,234 @@
+/*
```
Comment (elharo) [inline impl/maven-core/src/test/java/org/apache/maven/project/ProjectBuildingExceptionMessageDemo.java]: no demo code please. use automated unit tests instead

File: impl/maven-core/src/test/java/org/apache/maven/project/ProjectBuildingExceptionTest.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+                null, null, createProblems(createProblem("Some error", ModelProblem.Severity.ERROR, "unknown", 1, 1)));
+        results.add(result);
+
+        ProjectBuildingException exception = new ProjectBuildingException(results);
+        String message = exception.getMessage();
+
+        assertTrue(message.contains("[unknown project]"), "Message should handle unknown project gracefully");
+    }
+
+    private List<ModelProblem> createProblems(ModelProblem... problems) {
```
Comment (elharo) [inline impl/maven-core/src/test/java/org/apache/maven/project/ProjectBuildingExceptionTest.java]: These methods are simple enough to be inlined

**Review by elharo — CHANGES_REQUESTED** [review]

File: impl/maven-core/src/main/java/org/apache/maven/project/ProjectBuildingException.java [hunk 3]
```diff
@@ -97,4 +97,84 @@ private static String createMessage(String message, String projectId, File pomFi
...
     }
+
+    @SuppressWarnings("deprecation")
+    private static String createMessage(List<ProjectBuildingResult> results) {
+        if (results == null || results.isEmpty()) {
+            return "Some problems were encountered while processing the POMs";
+        }
+
+        long totalProblems = results.stream()
+                .filter(r -> r.getProblems() != null)
```
Comment (elharo) [inline impl/maven-core/src/main/java/org/apache/maven/project/ProjectBuildingException.java]: can this be null?

File: impl/maven-core/src/main/java/org/apache/maven/project/ProjectBuildingException.java [hunk 4]
```diff
@@ -97,4 +97,84 @@ private static String createMessage(String message, String projectId, File pomFi
...
+                .mapToLong(r -> r.getProblems().size())
+                .sum();
+
+        long errorProblems = results.stream()
+                .filter(r -> r.getProblems() != null)
```
Comment (elharo) [inline impl/maven-core/src/main/java/org/apache/maven/project/ProjectBuildingException.java]: can this be null?

File: impl/maven-core/src/main/java/org/apache/maven/project/ProjectBuildingException.java (hunk 3 above)
Comment (PasanAbeysekara) [reply impl/maven-core/src/main/java/org/apache/maven/project/ProjectBuildingException.java]: No, it cannot be null. Removed null handle as follows

`long totalProblems =
                results.stream().mapToLong(r -> r.getProblems().size()).sum();`

File: impl/maven-core/src/main/java/org/apache/maven/project/ProjectBuildingException.java (hunk 4 above)
Comment (PasanAbeysekara) [reply impl/maven-core/src/main/java/org/apache/maven/project/ProjectBuildingException.java]: No, it cannot be null. Removed null handle as follows

long errorProblems = results.stream()
                .flatMap(r -> r.getProblems().stream())
                .filter(p -> p.getSeverity() != org.apache.maven.model.building.ModelProblem.Severity.WARNING)
                .count();
